binderhub/repoproviders.py

```diff
@@ -111,6 +121,21 @@ def has_higher_quota(self):
                 return True
         return False
 
+    def spec_configuration_override(self):
+        """
+        Return the spec configuration overrides if the given spec has them.
+        """
+        if not isinstance(self.spec_configuration_override, dict):
+            raise ValueError(
+                "Specification-specific override is not a "
+                "dictionary, found type %s" % type(self.spec_configuration_override))
+        for spec, config in self.spec_configuration_override:
+            # Ignore case, because most git providers do not
+            # count DS-100/textbook as different from ds-100/textbook
+            if re.match(spec, self.spec, re.IGNORECASE):
```

Comment:
FWIW, using `re.match` compiles the regex on each call, which means we are compiling every regex here on every repo provider. Compiling them on application load would make this a bit more efficient.

Related to this: dicts are in a random order, but overrides might want to have a priority if more than one pattern matches the repo. With a dict, the behavior is undefined, with a list it would be consistent. So perhaps a list of tuples is more appropriate than a dict?

Reply:
mmmm, I am not sure how to do this :-) feel free to suggest the code that would still allow for config patterns like the one in https://github.com/jupyterhub/binderhub/pull/888#issue-294300289 and I'll give it a shot

Reply:
Good point about the random order! I had pondered that one might have multiple patterns and if they should be additive or "last one wins" or what, then decided that for a first pass we should ignore that. However the random order thing is something we need to address (or inform the user that they have more than one pattern that matches and that this is an error (for now)).

I think what we want is a list like:
```python
specs = [{"pattern": "some-pattern", "quota": 10}, {"pattern": "some-other-pattern", "quota": 12}]
```
in YAML it would look like:
```yaml
# not super sure about the first one but the second two should work
- {pattern: "yet another pattern", quota: 33}
- pattern: some-pattern
  quota: 10
- pattern: some-other-pattern
  quota: 12
```

Reply:
Ah I like this pattern! @minrk does it look good to you as well?

Reply:
yes, that looks great to me! The only question is that 'pattern' is in the same namespace as the config overrides. This is nice since it's more concise, but could allow collisions or confusion. The more rigorous, but slightly more tedious would be to explicitly separate the match from the override config:

```yaml
- pattern: some-pattern
  config:
    quota: 10
    other_config: "x"
```

This would allow adding other siblings to `pattern` if we e.g. had some other options that should influence the selection rather than the overridden config. I don't have strong feelings, but I've occasionally regretted not separating things like this in JupyterHub in the past (see spawner options in the REST API).

You can see the difference in code where 'pattern' needs to be handled specially:

```python
# copy because we need to modify the dict
update_config = config.copy()
# remove pattern before updating to config because pattern is not part of the config:
pattern = config.pop('pattern')
if pattern.match(...):
    config.update(update_config)
```

vs:

```python
pattern = item['pattern']
update_config = item['config']
if pattern.match(spec):
    config.update(update_config)
```